outlines/generate/__init__.py

```diff
@@ -6,3 +12,45 @@
 from .json import json
 from .regex import regex
 from .text import text
+
+if TYPE_CHECKING:
+    from outlines.processors import RegexLogitsProcessor
+
+
+@dataclass
+class Generator:
+    """Represents a generator.
+
+    We use a class that can hold the logits processor which can be quite
+    expensive to build.
+
+    Attributes
+    ----------
+    model
+        An instance of a model wrapper.
+    output_description
+        The output type for API models, and the processor for other models.
+
+    """
+
+    model: Union[APIModel, Model]
+    output_description: Union["RegexLogitsProcessor", Any]
+
+    def __init__(
+        self, model, output_type: Optional[Union[Json, List, Choice, Regex]] = None
+    ):
+        from outlines.processors import RegexLogitsProcessor
+
+        if isinstance(model, get_args(APIModel)):
+            self.output_description = output_type
+        else:
+            if output_type is not None:
+                regex_string = output_type.to_regex()
+                self.output_description = RegexLogitsProcessor(
+                    regex_string, model.tokenizer
+                )
+            else:
+                self.output_description = None
```

Comment:
You only manipulating with `self.output_description` and don't assign `self.model = model`, which would fully redefine class attrs in `__init__` and could be okay, but for such partial redefinition in `@dataclass` case I see people suggest `__post_init__`, which will be called after `__init__`? But I'm not sure here.

Overall, took me a second here to understand this fork: `APIModel` <-> `output type`, `Model` <-> `processor`. I can't put my finger on it with a proper design idea, but the fact that we have two good matching pairs suggests a good potential for separation.

Reply:
Yeah I agree, maintain a single interface for API-based models and local models is annoying.

One possible way out of this design issue is to have models build and hold the `LogitsProcessor` but it does feel unintuitive. The other solution is to add a `Generator` building function that dispatches to an `APIGenerator` and a `LogitsProcessorGenerator`, but I feel the duplication of code is not necessarily justified. It might be more a case of documenting the behavior better?

I'll see if I can come up with something better.

Reply:
> The other solution is to add a Generator building function that dispatches to an APIGenerator and a LogitsProcessorGenerator, but I feel the duplication of code is not necessarily justified.

Actually, this might be enough, I'd vote for duplications like this to make design more self-descriptive rather than bringing it on a higher level of clarifications in docs

Reply:
I made the change. It actually does not look like duplication so much, and makes the code much easier to follow.

Reply:
It was worth it, crystal clear now 🔥